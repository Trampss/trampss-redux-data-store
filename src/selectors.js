import { at } from 'lodash'

export const getState = path => prefix => (state) => {
  let subState = state
  if (path !== undefined) subState = at(state, path)[0]

  return subState[prefix]
}

const getFactory = key => path => prefix => state => getState(path)(prefix)(state)[key]
export const getKeys = getFactory('keys')
export const getAsArray = getFactory('array')
export const getNb = getFactory('nb')
export const isInitialized = getFactory('initialized')

const getDatas = getFactory('datas')
export const get = path => prefix => state => (keys) => {
  const datas = getDatas(path)(prefix)(state)
  // All datas
  if (!keys) return datas
  // By keys
  if (Array.isArray(keys)) return keys.map(k => datas[k])
  // By key
  return datas[keys]
}

