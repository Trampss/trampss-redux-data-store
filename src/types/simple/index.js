import middleware from './simple.middleware'
import * as actions from './simple.actions'
import * as selectors from './simple.selectors'

export default {
  middlewares: [middleware],
  actions,
  selectors,
}
