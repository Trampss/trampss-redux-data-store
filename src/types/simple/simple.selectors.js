import { initState } from './simple.middleware'
import { getState } from '../../selectors'

export const get = options => () => state => getState(options)(state)

export const isInitialized = options => (state) => {
  if (options.defaultData !== undefined) return getState(options)(state) !== options.defaultData
  return get(options)()(state) !== initState
}
