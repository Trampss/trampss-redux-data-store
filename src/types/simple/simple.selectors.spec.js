/* eslint-env jest */
import * as selectors from './simple.selectors'
import { initState } from './simple.middleware'

describe('selectors/simple => ', () => {
  describe('isInitialized', () => {
    describe('without defaultData', () => {
      const options = { name: 'o' }
      it('should not be initialized', () => {
        expect(selectors.isInitialized(options)({ o: initState })).toMatchSnapshot()
      })
      it('should be initialized -object-', () => {
        expect(selectors.isInitialized(options)({ o: { i: 'initialized' } })).toMatchSnapshot()
      })
      it('should be initialized -string-', () => {
        expect(selectors.isInitialized(options)({ o: 'initialized' })).toMatchSnapshot()
      })
    })

    describe('defaultData as a string', () => {
      const options = { name: 'o', defaultData: 'defaultData' }
      it('should not be initialized', () => {
        expect(selectors.isInitialized(options)({ o: options.defaultData })).toMatchSnapshot()
      })
      it('should be initialized', () => {
        expect(selectors.isInitialized(options)({ o: 'initialized' })).toMatchSnapshot()
      })
    })

    describe('defaultData as an object', () => {
      const options = { name: 'o', defaultData: {} }
      it('should not be initialized', () => {
        expect(selectors.isInitialized(options)({ o: options.defaultData })).toMatchSnapshot()
      })
      it('should be initialized', () => {
        expect(selectors.isInitialized(options)({ o: { i: 'initialized' } })).toMatchSnapshot()
      })
    })
  })

  describe('get', () => {
    describe('without defaultData', () => {
      const options = { name: 'o' }
      it('should return the initState', () => {
        expect(selectors.get(options)()({ o: initState })).toMatchSnapshot()
      })
    })

    describe('defaultData as a string', () => {
      const options = { name: 'o', defaultData: '' }
      it('should return the default data', () => {
        expect(selectors.get(options)()({ o: options.defaultData })).toMatchSnapshot()
      })
      it('should return the current data', () => {
        expect(selectors.get(options)()({ o: 'DATA' })).toMatchSnapshot()
      })
    })

    describe('defaultData as an object', () => {
      const options = { name: 'o', defaultData: {} }
      it('should return the default data', () => {
        expect(selectors.get({ name: 'o' })()({ o: options.defaultData })).toMatchSnapshot()
      })
      it('should return the current data', () => {
        expect(selectors.get(options)()({ o: { i: 'DATA' } })).toMatchSnapshot()
      })
    })
  })
})
