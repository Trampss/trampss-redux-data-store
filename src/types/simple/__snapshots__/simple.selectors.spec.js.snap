// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`selectors/simple =>  get defaultData as a string should return the current data 1`] = `"DATA"`;

exports[`selectors/simple =>  get defaultData as a string should return the default data 1`] = `""`;

exports[`selectors/simple =>  get defaultData as an object should return the current data 1`] = `
Object {
  "i": "DATA",
}
`;

exports[`selectors/simple =>  get defaultData as an object should return the default data 1`] = `Object {}`;

exports[`selectors/simple =>  get without defaultData should return the initState 1`] = `Object {}`;

exports[`selectors/simple =>  isInitialized defaultData as a string should be initialized 1`] = `true`;

exports[`selectors/simple =>  isInitialized defaultData as a string should not be initialized 1`] = `false`;

exports[`selectors/simple =>  isInitialized defaultData as an object should be initialized 1`] = `true`;

exports[`selectors/simple =>  isInitialized defaultData as an object should not be initialized 1`] = `false`;

exports[`selectors/simple =>  isInitialized without defaultData should be initialized -object- 1`] = `true`;

exports[`selectors/simple =>  isInitialized without defaultData should be initialized -string- 1`] = `true`;

exports[`selectors/simple =>  isInitialized without defaultData should not be initialized 1`] = `false`;
